discovery-infra/bootstrap_in_place.py

```diff
@@ -0,0 +1,172 @@
+
+
+def str_presenter(dumper, data):
+    if "ssh-rsa" in data:  # check for multiline string
+        return dumper.represent_scalar('tag:yaml.org,2002:str', data, style='|')
+    return dumper.represent_scalar('tag:yaml.org,2002:str', data)
+
+
+def create_controller(net_asset):
+    return TerraformController(
+        cluster_name="test-infra-cluster",
+        num_masters=1,
+        num_workers=0,
+        master_memory=32 * 1024,  # 32GB of RAM
+        master_vcpu=12,
+        net_asset=net_asset,
+        iso_download_path="<TBD>",  # will be set later on
+        bootstrap_in_place=True,
+    )
+
+
+def all_operators_up():
+    statuses = get_operators_status(KUBE_CONFIG)
+    if not statuses:
+        logging.debug("No operator has been found currently...")
+        return False
+
+    invalid_operators = [operator for operator, up in statuses.items() if not up]
+
+    all_operators_are_valid = len(invalid_operators) == 0
+
+    if not all_operators_are_valid:
+        logging.debug("Following operators are still down: %s", ", ".join(invalid_operators))
+
+    return all_operators_are_valid
+
+
+def gather_sosreport_data(node):
+    node.upload_file(SOSREPORT_SCRIPT, "/tmp/man_sosreport.sh")
+    node.run_command("chmod a+x /tmp/man_sosreport.sh")
+    node.run_command("sudo /tmp/man_sosreport.sh")
+    node.download_file("/tmp/sosreport.tar.bz2", IBIP_DIR)
+
+
+def waiting_for_installation_completion(controller):
+    logging.info("Configuring /etc/hosts...")
+    utils.config_etc_hosts(cluster_name=controller.cluster_name,
+                           base_dns_domain=controller.cluster_domain,
+                           api_vip=controller.master_ips[0][0])
+
+    logging.info("Waiting for installation to complete...")
+    waiting.wait(all_operators_up,
+                 sleep_seconds=20,
+                 timeout_seconds=60 * 60,
+                 waiting_for="all operators to get up")
+    logging.info("Installation completed successfully!")
+
+
+def execute_ibip_flow(args):
+    openshift_release_image = os.getenv('OPENSHIFT_INSTALL_RELEASE_IMAGE')
+    if not openshift_release_image:
+        raise ValueError("os env OPENSHIFT_INSTALL_RELEASE_IMAGE must be provided")
+
+    net_asset = NetworkAssets().get()
+    controller = create_controller(net_asset)
+    setup_files_and_folders(args, net_asset, controller.cluster_name)
+
+    utils.extract_installer(openshift_release_image, BUILD_DIR)
+    installer_generate()
+
+    download_live_image(f"{BUILD_DIR}/installer-image.iso")
+    image_path = embed("installer-image.iso", "bootstrap.ign", EMBED_IMAGE_NAME)
+
+    logging.info("Starting node...")
+    controller.image_path = image_path
+    controller.start_all_nodes()
+    node = Nodes(controller, private_ssh_key_path=SSH_KEY)[0]
+    logging.info("Node started!")
+
+    try:
+        waiting_for_installation_completion(controller)
+    finally:
+        logging.info("Gathering sosreport data from host...")
```

Comment:
can we add this finally code to function too?

Reply:
done